agent/src/util/device_plugin_service.rs

```diff
@@ -455,962 +534,2299 @@ impl InstanceDevicePlugin {
-                    .await
+                if let Err(e) = try_update_instance_device_usage(
+                    &device_usage_id,
+                    &dps.node_name,
+                    &instance_name,
+                    &dps.config_namespace,
+                    DeviceUsageKind::Configuration(vdev_id.clone()),
+                    kube_interface.clone(),
+                )
+                .await
                 {
-                    if x == (MAX_INSTANCE_UPDATE_TRIES - 1) {
-                        trace!("try_update_instance_device_usage - update_instance returned error [{}] after max tries ... returning error", e);
-                        return Err(Status::new(Code::Unknown, "Could not update Instance"));
-                    }
-                    random_delay().await;
-                } else {
-                    return Ok(());
-                }
-            }
-            DeviceUsageStatus::ReservedByConfiguration(_) => {
-                if matches!(desired_device_usage_kind, DeviceUsageKind::Configuration(_)) {
-                    return Ok(());
-                } else {
-                    return Err(Status::new(
-                        Code::Unknown,
-                        "Requested device already in use",
-                    ));
-                }
-            }
-            DeviceUsageStatus::ReservedByInstance => {
-                if matches!(desired_device_usage_kind, DeviceUsageKind::Instance) {
-                    return Ok(());
-                } else {
-                    return Err(Status::new(
-                        Code::Unknown,
-                        "Requested device already in use",
-                    ));
+                    trace!("ConfigurationDevicePlugin::allocate - could not assign {} slot to {} node ... forcing list_and_watch to continue", device_usage_id, &dps.node_name);
+                    dps.list_and_watch_message_sender
+                        .send(ListAndWatchMessageKind::Continue)
+                        .unwrap();
+                    return Err(e);
                 }
-            }
-            DeviceUsageStatus::ReservedByOtherNode => {
-                trace!("try_update_instance_device_usage - request for device slot {} previously claimed by a diff node {} than this one {} ... indicates the device on THIS node must be marked unhealthy, invoking ListAndWatch ... returning failure, next scheduling should succeed!",
-                    device_usage_id, current_device_usage.get_node_name(), node_name);
-                return Err(Status::new(
-                    Code::Unknown,
-                    "Requested device already in use",
-                ));
-            }
-            DeviceUsageStatus::Unknown => {
+
+                let node_usage = NodeUsage::create(
+                    &DeviceUsageKind::Configuration(vdev_id.clone()),
+                    &dps.node_name,
+                )
+                .unwrap();
```

Comment:
i don't remember if there was some policy on when to return an error and when to unwrap, do we want to return an error here instead of unwrap call?

Reply:
mostly, i just want to make sure that unwarp is called in appropriate and intentional places vs returning an error.

Reply:
dps.list_and_watch_message_sender is for notifying list_and_watch thread within the same device plugin service to re-scan the device availability.  Since it is always valid as long as the device plugin service is running, the unwrap() should always succeed, if not, then it's unexpected and we fail fast.